File: webclient/src/sound/Envelope.ts
import Packet from '#/io/Packet.js';

export default class Envelope {
    start: number = 0;
    end: number = 0;
    form: number = 0;

    private envLength: number = 0;
    private shapeDelta: Int32Array | null = null;
    private shapePeak: Int32Array | null = null;
    private envThreshold: number = 0;
    private envPosition: number = 0;
    private delta: number = 0;
    private envAmplitude: number = 0;
    private ticks: number = 0;

    read(dat: Packet): void {
        this.form = dat.g1();
        this.start = dat.g4();
        this.end = dat.g4();
        this.envLength = dat.g1();
        this.shapeDelta = new Int32Array(this.envLength);
        this.shapePeak = new Int32Array(this.envLength);

        for (let i: number = 0; i < this.envLength; i++) {
            this.shapeDelta[i] = dat.g2();
            this.shapePeak[i] = dat.g2();
        }
    }

    reset(): void {
        this.envThreshold = 0;
        this.envPosition = 0;
        this.delta = 0;
        this.envAmplitude = 0;
        this.ticks = 0;
    }

    evaluateAt(delta: number): number {
        if (this.ticks >= this.envThreshold && this.shapePeak && this.shapeDelta) {
            this.envAmplitude = this.shapePeak[this.envPosition++] << 15;

            if (this.envPosition >= this.envLength) {
                this.envPosition = this.envLength - 1;
            }

            this.envThreshold = ((this.shapeDelta[this.envPosition] / 65536.0) * delta) | 0;
            if (this.envThreshold > this.ticks) {
                this.delta = (((this.shapePeak[this.envPosition] << 15) - this.envAmplitude) / (this.envThreshold - this.ticks)) | 0;
            }
        }

        this.envAmplitude += this.delta;
        this.ticks++;
        return (this.envAmplitude - this.delta) >> 15;
    }
}

File: javaclient/src/main/java/jagex2/sound/Envelope.java
package jagex2.sound;

import deob.ObfuscatedName;
import jagex2.io.Packet;

@ObfuscatedName("bc")
public class Envelope {

	@ObfuscatedName("bc.b")
	public int length;

	@ObfuscatedName("bc.c")
	public int[] shapeDelta;

	@ObfuscatedName("bc.d")
	public int[] shapePeak;

	@ObfuscatedName("bc.e")
	public int start;

	@ObfuscatedName("bc.f")
	public int end;

	@ObfuscatedName("bc.g")
	public int form;

	@ObfuscatedName("bc.h")
	public int threshold;

	@ObfuscatedName("bc.i")
	public int position;

	@ObfuscatedName("bc.j")
	public int delta;

	@ObfuscatedName("bc.k")
	public int amplitude;

	@ObfuscatedName("bc.l")
	public int ticks;

	@ObfuscatedName("bc.a(ZLmb;)V")
	public final void unpack(Packet arg1) {
		this.form = arg1.g1();
		this.start = arg1.g4();
		this.end = arg1.g4();
		this.length = arg1.g1();
		this.shapeDelta = new int[this.length];
		this.shapePeak = new int[this.length];
		for (int var3 = 0; var3 < this.length; var3++) {
			this.shapeDelta[var3] = arg1.g2();
			this.shapePeak[var3] = arg1.g2();
		}
	}

	@ObfuscatedName("bc.a(B)V")
	public final void genInit() {
		this.threshold = 0;
		this.position = 0;
		this.delta = 0;
		this.amplitude = 0;
		this.ticks = 0;
	}

	@ObfuscatedName("bc.a(II)I")
	public final int genNext(int arg1) {
		if (this.ticks >= this.threshold) {
			this.amplitude = this.shapePeak[this.position++] << 15;
			if (this.position >= this.length) {
				this.position = this.length - 1;
			}
			this.threshold = (int) ((double) this.shapeDelta[this.position] / 65536.0D * (double) arg1);
			if (this.threshold > this.ticks) {
				this.delta = ((this.shapePeak[this.position] << 15) - this.amplitude) / (this.threshold - this.ticks);
			}
		}
		this.amplitude += this.delta;
		this.ticks++;
		return this.amplitude - this.delta >> 15;
	}
}
